File: examples/deimv2/cpp/deimv2.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <iomanip>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

#include <opencv2/opencv.hpp>

#include "vision_service.h"

namespace {

void usage(const char* program) {
    std::cout
        << "Usage: " << program << " <config.yaml> [options]\n"
        << "  --model-path <path>  Override model_path\n"
        << "  --image <path>       Override test_image\n"
        << "  --output <path>      Output image (default: deimv2_result.jpg)\n";
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        usage(argv[0]);
        return 1;
    }
    const std::string config_path = argv[1];
    std::string model_path;
    std::string image_path;
    std::string output_path = "deimv2_result.jpg";
    for (int index = 2; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--model-path" && index + 1 < argc) {
            model_path = argv[++index];
        } else if (argument == "--image" && index + 1 < argc) {
            image_path = argv[++index];
        } else if (argument == "--output" && index + 1 < argc) {
            output_path = argv[++index];
        } else if (argument == "--help") {
            usage(argv[0]);
            return 0;
        } else {
            std::cerr << "Unknown or incomplete argument: "
                << argument << std::endl;
            return 1;
        }
    }

    std::unique_ptr<VisionService> service =
        VisionService::Create(config_path, model_path, false);
    if (!service) {
        std::cerr << "Create failed: "
            << VisionService::LastCreateError() << std::endl;
        return 1;
    }
    if (image_path.empty()) {
        image_path = service->GetDefaultImage();
    }
    cv::Mat image = cv::imread(image_path);
    if (image.empty()) {
        std::cerr << "Could not read image: " << image_path << std::endl;
        return 1;
    }

    VisionServiceResponse response;
    const VisionServiceStatus status = service->Infer(image, &response);
    if (status != VISION_SERVICE_OK || !response.ok) {
        std::cerr << "Infer failed: " << service->LastError() << std::endl;
        return 1;
    }

    const std::vector<std::string> class_names =
        service->GetClassNames();
    std::cout << "Detections: " << response.results.size() << std::endl;
    for (const vision::Result& result : response.results) {
        const int label = vision::get_label(result);
        const vision::BoundingBox box = vision::get_bbox(result);
        const std::string name =
            label >= 0 && static_cast<size_t>(label) < class_names.size()
            ? class_names[static_cast<size_t>(label)]
            : "class " + std::to_string(label);
        std::cout << "  " << name << " (" << label << "), score: "
            << std::fixed << std::setprecision(4)
            << vision::get_score(result) << ", box: ["
            << box.x1 << ", " << box.y1 << ", "
            << box.x2 << ", " << box.y2 << "]\n";
    }
    if (!service->SupportsDraw()) {
        std::cerr << "DEIMv2 must support Draw" << std::endl;
        return 1;
    }
    cv::Mat output;
    if (service->Draw(image, response, &output) != VISION_SERVICE_OK ||
        output.empty()) {
        std::cerr << "Draw failed: " << service->LastError() << std::endl;
        return 1;
    }
    if (!cv::imwrite(output_path, output)) {
        std::cerr << "Could not write output: " << output_path << std::endl;
        return 1;
    }
    std::cout << "Saved: " << output_path << std::endl;
    return 0;
}

File: examples/deimv2/python/deimv2.py
#!/usr/bin/env python3
# Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
# SPDX-License-Identifier: Apache-2.0

"""DEIMv2-N object detection through the unified VisionService binding."""

import argparse
from pathlib import Path

import cv2

from spacemit_vision import VisionServiceNative, VisionServiceStatus


def parse_args():
    parser = argparse.ArgumentParser(description="DEIMv2-N example")
    parser.add_argument(
        "--config",
        default=str(Path(__file__).parent.parent / "config" / "deimv2.yaml"),
    )
    parser.add_argument("--model-path", default="")
    parser.add_argument("--image")
    parser.add_argument("--output", default="deimv2_result.jpg")
    parser.add_argument("--conf-threshold", type=float, default=-1.0)
    return parser.parse_args()


def main():
    args = parse_args()
    service = VisionServiceNative.create(
        args.config, model_path_override=args.model_path
    )
    try:
        image_path = args.image or service.get_default_image()
        if not image_path:
            raise ValueError("No image; use --image or set test_image in config")
        image = cv2.imread(str(Path(image_path).expanduser()))
        if image is None:
            raise FileNotFoundError(f"Could not read image: {image_path}")

        status, results = service.infer_image(
            image, conf=args.conf_threshold, iou=-1.0
        )
        if status != VisionServiceStatus.OK:
            raise RuntimeError(service.last_error())

        class_names = service.get_class_names()
        print(f"Detections: {len(results)}")
        for result in results:
            name = (
                class_names[result.label]
                if 0 <= result.label < len(class_names)
                else f"class {result.label}"
            )
            print(
                f"  {name} ({result.label}), score: {result.score:.4f}, "
                f"box: [{result.x1:.2f}, {result.y1:.2f}, "
                f"{result.x2:.2f}, {result.y2:.2f}]"
            )

        if not service.supports_draw():
            raise RuntimeError("DEIMv2 must support draw")
        draw_status, output = service.draw(image)
        if draw_status != VisionServiceStatus.OK or output is None:
            raise RuntimeError(service.last_error())
        if not cv2.imwrite(args.output, output):
            raise RuntimeError(f"Could not write output: {args.output}")
        print(f"Saved: {args.output}")
        return 0
    finally:
        service.release()


if __name__ == "__main__":
    raise SystemExit(main())
